packages/jupyter-ai/jupyter_ai/llm_handler.py

```diff
@@ -0,0 +1,70 @@
+from typing import Dict, Type
+
+from jupyter_ai.config_manager import ConfigManager, Logger
+from jupyter_ai_magics.providers import BaseProvider
+
+
+class BaseLLMHandler:
+    """Base class containing shared methods and attributes used by LLM handler classes."""
+
+    handler_kind: str
+
+    def __init__(
+        self,
+        log: Logger,
+        config_manager: ConfigManager,
+        model_parameters: Dict[str, Dict],
+    ):
+        self.log = log
+        self.config_manager = config_manager
+        self.model_parameters = model_parameters
+        self.llm = None
+        self.llm_params = None
+        self.llm_chain = None
+
+    def model_changed_callback(self):
+        """Method which can be overridden in sub-classes to listen to model change."""
+        pass
+
+    def get_llm_chain(self):
+        lm_provider = self.config_manager.lm_provider
+        lm_provider_params = self.config_manager.lm_provider_params
+
```

Comment:
I would like to make the completion model/provider separately configurable from the chat model. I could make this mixin class configurable with:

```python
lm_provider_key = 'lm_provider' # default for chat model handler
lm_provider_params_key = 'lm_provider_params' # default for chat model handler
```

And then for inline completion handler:

```python
lm_provider_key = 'lm_completion_provider'
lm_provider_params_key = 'lm_completion_provider_params'
```

Does it sound ok? As this might increase the changeset considerably (and the number of conflicts) I would consider doing this in follow-up PR, depending on what is your availability for review/merging this one.

Reply:
See my main review comment above. I don't think we should be overly eager on sharing the code contained in this class between other classes.